### PR TITLE
[db] Fix string booleans to be booleans

### LINKED ISSUE #5328: Change existing string values in DB to booleans
### Bug description

Screenshot 2021-08-23 at 09 15 25

Some values of allowsChangelogMail are strings when they should be booleans. This was caused by a migration on 20.08.2021 where the query incorrectly populated the field as strings, therefore it only affected data up to that date.

### Steps to reproduce

1. Connect to the Gitpod DB in production.
2. Run `select name, additionalData from from d_b_user`.

### Expected behavior

"allowsChangelogMail" should be either `true` or `false`.

### Example repository

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by laushinka**
Fixes #5328

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **laushinka** after the PR has been reviewed.

Associated issue: *#5328*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by JanKoehnlein** [comment]
/cc @AlexTugarev
/cc @jakobhero 
/uncc @atduarte 

I am not sure if this migration is worthwhile. WDYT?

**Comment by JanKoehnlein** [comment]
/hold 

Let's not jeopardize today's deployment. If reviewers are sure it is harmless, unhold.

**Comment by laushinka** [comment]
@jakobhero @JanKoehnlein @AlexTugarev Are we still going ahead with this PR?

**Comment by JanKoehnlein** [comment]
It mostly affects downstream usages, so maybe @jakobhero should decide whether that's an improvement.

**Comment by jakobhero** [comment]
i think this is an improvement as it removes the inconsistencies of date formats in the additional data field, which at the moment a)  is a potential source for errors in the if someone queries the data in the app in the future and b) requires cleanup on the analytics side. while i can live with b), i think it should be preferrable to have a clean data source in the first place.
this looks good to me!

**Review by jakobhero — APPROVED** [review]

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: c1c6714530919d1e84ca8404388890a6807571c0</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/5439#pullrequestreview-747184295" title="Approved">jakobhero</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **laushinka** after the PR has been reviewed.

Associated issue: *#5328*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by laushinka** [comment]
> It mostly affects downstream usages, so maybe @jakobhero should decide whether that's an improvement.

@JanKoehnlein Yup, this change was requested by @jakobhero.

**Comment by JanKoehnlein** [comment]
OK, let's do it after tomorrow's (7th Sept) deployment.

**Comment by laushinka** [comment]
> OK, let's do it after tomorrow's (7th Sept) deployment.

@JanKoehnlein Do we have a deployment tomorrow? I thought today's replaced tomorrow's.

File: components/gitpod-db/src/typeorm/migration/1630333424897-ReplaceStringBooleansWithBooleans.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import {MigrationInterface, QueryRunner} from "typeorm";
+import { columnExists } from "./helper/helper";
+
+export class ReplaceStringBooleansWithBooleans1630333424897 implements MigrationInterface {
+
+    public async up(queryRunner: QueryRunner): Promise<any> {
+        if (await columnExists(queryRunner, "d_b_user", "additionalData")) {
+            await queryRunner.query("update d_b_user set additionalData = json_set(additionalData, '$.emailNotificationSettings.allowsChangelogMail', true) where json_extract(additionalData, '$.emailNotificationSettings.allowsChangelogMail') = 'true'");
+            await queryRunner.query("update d_b_user set additionalData = json_set(additionalData, '$.emailNotificationSettings.allowsChangelogMail', false) where json_extract(additionalData, '$.emailNotificationSettings.allowsChangelogMail') = 'false'")
```
Comment (AlexTugarev) [inline components/gitpod-db/src/typeorm/migration/1630333424897-ReplaceStringBooleansWithBooleans.ts:15]: I'm not very familiar with json functions in mysql, thus had to look it up first. ;-)

wdyt of using `JSON_UNQUOTE` and generalize, which would shave off the second query. I mean something like this:

```
son_set(additionalData, '$.emailNotificationSettings.allowsChangelogMail', JSON_UNQUOTE(json_extract(additionalData, '$.emailNotificationSettings.allowsChangelogMail')))
```

**Comment by AlexTugarev** [comment]
@laushinka and @jakobhero, does this change, which looks good in general, fix anything user facing? I'm asking because it would be great to avoid any blocking queries on the user table during deployment, as this may indeed have impact on UX. Unfortunately, we didn't record any statistics on db migration, did we?

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/5439#issuecomment-916835180" title="LGTM">AlexTugarev</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/5439#pullrequestreview-747184295" title="Approved">jakobhero</a>*

Associated issue: *#5328*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [AlexTugarev]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/gitpod-db/src/typeorm/migration/1630333424897-ReplaceStringBooleansWithBooleans.ts (hunk 1 above)
Comment (laushinka) [reply components/gitpod-db/src/typeorm/migration/1630333424897-ReplaceStringBooleansWithBooleans.ts:15]: Nice find with the JSON_UNQUOTE! Let's see what the DB expert @jakobhero says.

Comment (jakobhero) [reply components/gitpod-db/src/typeorm/migration/1630333424897-ReplaceStringBooleansWithBooleans.ts:15]: i like the idea of the generalizing, but `JSON_QUOTE()` returns a string and the purpose of the migration is to use booleans as the consistent data type across all email notification settings. having a general solution that would use booleans could be achieved 
by checking for equality with `'true'` (cp. attached screenshot):
```
json_set(additionalData, '$.emailNotificationSettings.allowsChangelogMail', JSON_UNQUOTE(json_extract(additionalData, '$.emailNotificationSettings.allowsChangelogMail'))='true')
```
Screenshot 2021-09-13 at 19 57 24

**Comment by jakobhero** [comment]
> does this change fix anything user facing?
it does not have any effect on user, the purpose of the PR is to have a consistent data format in the DB to make it easier to work with and a source for future errors.

**Pull request merged**